crma_api_client/client.py

```diff
@@ -0,0 +1,184 @@
+            version: CRMA REST API version
+            timeout: Default timeout for requests and non-connect operations, in seconds
+            connect_timeout: Default timeout for establishing an HTTP connection, in
+                seconds
+            logger: Custom logger instance to use instead of the stdlib
+
+        """
+        self.logger = logger
+        self._client = httpx.AsyncClient(
+            base_url=conn.instance_url.rstrip("/") + f"/services/data/{version}",
+            headers={"Authorization": conn.authorization},
+            timeout=httpx.Timeout(timeout, connect=connect_timeout),
+        )
+
+    async def _get_headers(
+        self,
+    ) -> Dict[str, str]:
+        """Get a headers dict from the calling context
+
+        Returns:
+            Dict containing headers to include in requests
+
+        """
+        headers = {
+            "content-type": "application/json",
+            "accept": "application/json",
+        }
+
+        return headers
+
+    @backoff.on_exception(
+        backoff.expo,
+        httpx.HTTPStatusError,
+        max_tries=3,
+        giveup=lambda e: e.response.status_code != 502,
+    )
+    async def request(
+        self,
+        path: str,
+        method: str,
+        json_data: Optional[Any] = None,
+        params: Optional[Dict[str, Any]] = None,
+        **kwargs: Any,
+    ) -> httpx.Response:
+        """Generic method to send a JSON request to the service
+
+        Args:
+            path: Path to the API resource. This path will be appended to the base url.
+            method: HTTP method
+            json_data: Request payload (for POST/PUT/PATCH requests)
+            params: Request query params
+
+        Returns:
+            response object
+
+        """
+        path = "/" + path.strip("/")
+        if json_data:
+            json_data = json_dumps_common(json_data).encode()
+        headers = await self._get_headers()
+        self.logger.debug(f"Service request starting path={path} method={method}")
+        response = await self._client.request(
+            method.upper(),
+            path,
+            headers=headers,
+            content=json_data,
+            params=params,
+            **kwargs,
+        )
+        self.logger.debug(
+            f"Service request completed status_code={response.status_code}"
+        )
+        response.raise_for_status()
+        return response
+
+    async def list_dataset_versions(self, identifier: str) -> DatasetVersionsResponse:
+        """List the versions for a dataset
+
+        Args:
+            identifier: Dataset name or ID
+
+        Returns:
+            DatasetVersionsResponse object
+
+        """
+        response = await self.request(f"/wave/datasets/{identifier}/versions", "GET")
+        return DatasetVersionsResponse.parse_obj(response.json())
+
+    async def query(self, req: QueryRequest) -> QueryResponse:
```

Comment:
I don't feel strongly, but it seems like this would be easier to read/use if the request arguments were defined inline instead of via the `QueryRequest` object?